imports/client/components/FancyEditor.tsx

```diff
@@ -0,0 +1,639 @@
+      if (token.type === 'link') {
+        ranges.push({
+          link: { href: token.href },
+          anchor: { path, offset },
+          focus: { path, offset: offset + token.raw.length },
+        });
+      } else if (token.type === 'blockquote') {
+        let distance = token.raw.length;
+        if (token.raw.endsWith('\n\n')) {
+          distance -= 1;
+        }
+        ranges.push({
+          blockquote: true,
+          anchor: { path, offset },
+          focus: { path, offset: offset + distance },
+        });
+      } else if (token.type === 'strong') {
+        let type = 'strong';
+        if (token.raw.startsWith('__')) {
+          // Discord treats double-underline as underline, rather than strong.
+          // We'll do likewise.
+          type = 'underline';
+        }
+        ranges.push({
+          [type]: true,
+          anchor: { path, offset },
+          focus: { path, offset: offset + token.raw.length },
+        });
+      } else if (token.type === 'code') {
+        let distance = token.raw.length;
+        if (token.raw.endsWith('\n')) {
+          // Avoid including a trailing newline in the code span decoration;
+          // it makes the <code> block's background appear to extend too far.
+          distance -= 1;
+        }
+        ranges.push({
+          code: true,
+          anchor: { path, offset },
+          focus: { path, offset: offset + distance },
+        });
+      } else if (token.type === 'em' || token.type === 'codespan' || token.type === 'del') {
+        ranges.push({
+          [token.type]: true,
+          anchor: { path, offset },
+          focus: { path, offset: offset + token.raw.length },
+        });
+      }
+    }
+  }, 0);
+
+  if (DEBUG_EDITOR) {
+    // eslint-disable-next-line no-console
+    console.log('decorated', ranges);
+  }
+  return ranges;
+};
+
+const Portal = ({ children }: { children: React.ReactNode }) => {
+  return createPortal(children, document.body);
+};
+
+export interface FancyEditorHandle {
+  clearInput: () => void;
+}
+
+const FancyEditor = React.forwardRef(({
+  className, initialContent, placeholder, users, onContentChange, onSubmit,
+}: {
+  className?: string,
+  initialContent: Descendant[],
+  placeholder?: string,
+  users: Meteor.User[],
+  onContentChange: (content: Descendant[]) => void,
+  onSubmit: () => void,
+}, forwardedRef: React.Ref<FancyEditorHandle>) => {
+  const [editor] = useState(() => {
+    const upstreamEditor = withReact(withHistory(createEditor()));
+    return withSingleMessage(withMentions(upstreamEditor));
+  });
+
+  // The floating autocomplete box for @-mentions
+  const ref = useRef<HTMLDivElement>(null);
+  // selection target?
+  const [target, setTarget] = useState<Range | undefined>(undefined);
+  // Offset into the list of potentially-matching usernames
+  const [index, setIndex] = useState<number>(0);
+  // current needle for search in display names
+  const [search, setSearch] = useState('');
```

Comment:
should these have names that incorporate "autocompletion" or something?

Reply:
Sure, these can be described a bit better.  Addressed in 62f0a8803e341e77b4579a8155455b2fb566b657